src/coreclr/jit/morphblock.cpp

```diff
@@ -1031,6 +1006,40 @@ void MorphCopyBlockHelper::MorphStructCases()
         m_result                 = CopyFieldByField();
         m_transformationDecision = BlockTransformation::FieldByField;
     }
+
+    // Mark the dest/src structs as DoNotEnreg when they are not being fully referenced as the same type.
+    //
+    if (!m_dstDoFldAsg && (m_dstVarDsc != nullptr) && !m_dstSingleLclVarAsg)
```

Comment:
after this change init/copy like:

               [000015] -ACXG-------              *  ASG       struct (copy)
               [000013] D----+-N----              +--*  LCL_VAR   struct<S, 8> V01 loc0
               [000012] --CXG+------              \--*  CALL      struct TestStructFields.Program.GetS1
don't require doNotEnreg flag. The condition was moved after we do m_comp->fgMorphBlockOperand to work after OBJ(ADDR(LCL)) collapse.